File: go/rrule/format.go
// SPDX-License-Identifier: MIT

package rrule

import (
	"strconv"
	"strings"

	vstar "hop.top/vstar"
)

// String renders the Rule as an RFC 5545 §3.3.10 RRULE property
// value with no "RRULE:" prefix, e.g.
// "FREQ=WEEKLY;INTERVAL=2;BYDAY=MO,WE".
//
// # Rule-part order
//
// RFC 5545 §3.3.10 imposes NO order on rule-parts — a producer may
// emit them any way it likes and the value means the same thing.
// spec/03 §RRULE wire form does impose one on emitters, so that
// identical logical content produces byte-identical output. String
// follows it, and the order is part of the API contract, not an
// implementation detail:
//
//	FREQ, INTERVAL, UNTIL, COUNT, BYMONTH, BYWEEKNO, BYYEARDAY,
//	BYMONTHDAY, BYDAY, BYHOUR, BYMINUTE, BYSECOND, BYSETPOS, WKST
//
// The order is the one RFC 5545 §3.3.10's own `recur` ABNF lists
// the rule-parts in, which makes it the least surprising choice
// and matches what most producers already emit. Reading it
// left-to-right also tracks decreasing period length — FREQ and
// its modifiers, then the termination bound, then the BY-* filters
// from coarsest (month) to finest (second) — with BYSETPOS last
// because it is applied last, as a positional filter over the
// fully expanded set, and WKST last of all because it is a
// modifier on the whole rule rather than a filter.
//
// # Defaults
//
// Rule-parts equal to their RFC default are omitted: INTERVAL=1
// and WKST=MO (spec/03 §RRULE wire form). This keeps the wire form
// minimal and means two Rules that differ only in whether the
// producer spelled out a default render identically.
//
// List values keep the order the Rule holds them in — RFC 5545
// gives BY-* lists no ordering semantics, and preserving producer
// order round-trips faithfully through ParseRRule. Callers wanting
// order-insensitive equality should compare parsed Rules, not
// strings.
//
// String returns "" for a Rule that cannot produce a valid RRULE
// value (FREQ unset), rather than emitting a partial value that
// would fail to re-parse.
func (r Rule) String() string {
	if r.Freq == FreqInvalid {
		return ""
	}

	var b strings.Builder
	b.WriteString("FREQ=")
	b.WriteString(r.Freq.String())

	// INTERVAL — omitted at its RFC default of 1.
	if r.Interval > 1 {
		b.WriteString(";INTERVAL=")
		b.WriteString(strconv.Itoa(r.Interval))
	}
	// UNTIL and COUNT are mutually exclusive per RFC 5545; emit
	// whichever is set.
	if !r.Until.IsZero() {
		b.WriteString(";UNTIL=")
		b.WriteString(vstar.FormatTime(r.Until))
	}
	if r.Count > 0 {
		b.WriteString(";COUNT=")
		b.WriteString(strconv.Itoa(r.Count))
	}

	writeIntList(&b, "BYMONTH", r.ByMonth)
	writeIntList(&b, "BYWEEKNO", r.ByWeekNo)
	writeIntList(&b, "BYYEARDAY", r.ByYearDay)
	writeIntList(&b, "BYMONTHDAY", r.ByMonthDay)
	writeByDayList(&b, r.ByDay)
	writeIntList(&b, "BYHOUR", r.ByHour)
	writeIntList(&b, "BYMINUTE", r.ByMinute)
	writeIntList(&b, "BYSECOND", r.BySecond)
	writeIntList(&b, "BYSETPOS", r.BySetPos)

	// WKST — omitted at its RFC default of MO.
	if r.WeekStart != MO {
		b.WriteString(";WKST=")
		b.WriteString(r.WeekStart.String())
	}

	return b.String()
}

// writeIntList appends ";NAME=v1,v2,..." when the list is
// non-empty, and writes nothing otherwise.
func writeIntList(b *strings.Builder, name string, vals []int) {
	if len(vals) == 0 {
		return
	}
	b.WriteByte(';')
	b.WriteString(name)
	b.WriteByte('=')
	for i, v := range vals {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteString(strconv.Itoa(v))
	}
}

// writeByDayList appends ";BYDAY=[<ordinal>]<weekday>,..." when the
// list is non-empty. A zero ordinal means "every weekday of this
// kind" and is rendered with no numeric prefix (per RFC 5545, the
// explicit "0" prefix is invalid).
func writeByDayList(b *strings.Builder, vals []ByDay) {
	if len(vals) == 0 {
		return
	}
	b.WriteString(";BYDAY=")
	for i, bd := range vals {
		if i > 0 {
			b.WriteByte(',')
		}
		if bd.Ordinal != 0 {
			b.WriteString(strconv.Itoa(bd.Ordinal))
		}
		b.WriteString(bd.Weekday.String())
	}
}

// Property renders the Rule as a complete vstar.Property ready to
// attach to a Component. Returns the zero Property when the Rule
// cannot produce a valid value (see String).
func (r Rule) Property() vstar.Property {
	v := r.String()
	if v == "" {
		return vstar.Property{}
	}
	return vstar.Property{Name: "RRULE", Value: v}
}

File: rs/src/rrule/format.rs
// SPDX-License-Identifier: MIT

//! The RRULE wire form.
//!
//! Canonical form preserves an `RRULE` value byte-for-byte (spec rule
//! 8), so this is the emitter's contract, not the canonical layer's:
//! two rules with the same logical content produce the same bytes only
//! because the emitter fixes an order and elides the defaults.
//!
//! The order is RFC 5545 §3.3.10's own `recur` ABNF order:
//!
//! ```text
//! FREQ, INTERVAL, UNTIL, COUNT, BYMONTH, BYWEEKNO, BYYEARDAY,
//! BYMONTHDAY, BYDAY, BYHOUR, BYMINUTE, BYSECOND, BYSETPOS, WKST
//! ```
//!
//! `INTERVAL=1` and `WKST=MO` are the RFC defaults and are omitted.
//!
//! **List values keep their authored order.** RFC 5545 gives `BY-*`
//! lists no ordering semantics, so sorting them would change bytes the
//! producer chose and break the round-trip. Callers wanting
//! order-insensitive equality compare parsed [`Rule`]s, not strings.

use std::fmt;

use crate::rrule::types::{ByDay, Rule, Weekday};
use crate::{format_time, Property};

impl fmt::Display for Rule {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "FREQ={}", self.freq)?;

        // INTERVAL — elided at its RFC default of 1.
        if self.interval > 1 {
            write!(f, ";INTERVAL={}", self.interval)?;
        }
        // UNTIL and COUNT are mutually exclusive; emit whichever is set.
        if let Some(until) = self.until {
            write!(f, ";UNTIL={}", format_time(until))?;
        }
        if let Some(count) = self.count {
            write!(f, ";COUNT={count}")?;
        }

        write_int_list(f, "BYMONTH", &self.by_month)?;
        write_int_list(f, "BYWEEKNO", &self.by_week_no)?;
        write_int_list(f, "BYYEARDAY", &self.by_year_day)?;
        write_int_list(f, "BYMONTHDAY", &self.by_month_day)?;
        write_by_day_list(f, &self.by_day)?;
        write_int_list(f, "BYHOUR", &self.by_hour)?;
        write_int_list(f, "BYMINUTE", &self.by_minute)?;
        write_int_list(f, "BYSECOND", &self.by_second)?;
        write_int_list(f, "BYSETPOS", &self.by_set_pos)?;

        // WKST — elided at its RFC default of MO.
        if self.week_start != Weekday::Mo {
            write!(f, ";WKST={}", self.week_start)?;
        }
        Ok(())
    }
}

impl Rule {
    /// The rule as a complete `RRULE` [`Property`], ready to attach to a
    /// component.
    ///
    /// ```
    /// use hop_top_vstar::rrule::parse_rrule;
    ///
    /// let p = parse_rrule("FREQ=DAILY;INTERVAL=2")?.to_property();
    /// assert_eq!(p.name, "RRULE");
    /// assert_eq!(p.value, "FREQ=DAILY;INTERVAL=2");
    /// assert!(p.params.is_empty());
    /// # Ok::<(), hop_top_vstar::Error>(())
    /// ```
    pub fn to_property(&self) -> Property {
        Property::new("RRULE", self.to_string())
    }
}

/// Writes `;NAME=v1,v2` for a non-empty list, in the order held.
fn write_int_list(f: &mut fmt::Formatter<'_>, name: &str, vals: &[i32]) -> fmt::Result {
    if vals.is_empty() {
        return Ok(());
    }
    write!(f, ";{name}=")?;
    for (i, v) in vals.iter().enumerate() {
        if i > 0 {
            f.write_str(",")?;
        }
        write!(f, "{v}")?;
    }
    Ok(())
}

/// Writes `;BYDAY=[<ordinal>]<weekday>,…`. A zero ordinal renders with
/// no prefix — the explicit `0` prefix is invalid per RFC 5545.
fn write_by_day_list(f: &mut fmt::Formatter<'_>, vals: &[ByDay]) -> fmt::Result {
    if vals.is_empty() {
        return Ok(());
    }
    f.write_str(";BYDAY=")?;
    for (i, bd) in vals.iter().enumerate() {
        if i > 0 {
            f.write_str(",")?;
        }
        write!(f, "{bd}")?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::rrule::parse::parse_rrule;
    use crate::rrule::types::{Freq, Rule, Weekday};

    #[test]
    fn elides_the_rfc_defaults() {
        assert_eq!(Rule::new(Freq::Daily).to_string(), "FREQ=DAILY");
        let mut r = Rule::new(Freq::Daily);
        r.interval = 2;
        r.week_start = Weekday::Su;
        assert_eq!(r.to_string(), "FREQ=DAILY;INTERVAL=2;WKST=SU");
    }

    #[test]
    fn keeps_list_order_as_authored() {
        let r = parse_rrule("FREQ=WEEKLY;BYDAY=SA,MO,WE").expect("parses");
        assert_eq!(r.to_string(), "FREQ=WEEKLY;BYDAY=SA,MO,WE");
    }

    #[test]
    fn re_emitting_is_idempotent() {
        for src in [
            "FREQ=DAILY",
            "FREQ=WEEKLY;INTERVAL=2;BYDAY=MO,WE",
            "FREQ=MONTHLY;BYMONTHDAY=-1;WKST=SU",
            "FREQ=DAILY;UNTIL=20260401T120000Z",
        ] {
            let once = parse_rrule(src).expect("parses").to_string();
            let twice = parse_rrule(&once).expect("re-parses").to_string();
            assert_eq!(once, twice, "{src}");
        }
    }
}
